.gitlab-ci.yml

```diff
@@ -1,14 +1,108 @@
-test:
-  image: "python:3.6-stretch"
-  script:
-    - apt-get update -qq
-    - apt-get install -y git
-    - apt-get install -y xsltproc
-    - apt-get install -y genisoimage
-    - apt-get install -y enchant
-    - apt-get install -y shellcheck
-    - apt-get install -y python3-pip
-    - pip3 install --upgrade pip
-    - pip install tox
+stages:
+  - test
+  - doc
+  - package
+  - deploy
+
+.fedora_install_deps: &fedora_install_deps
+  before_script:
+    - dnf install -y --refresh python3 python3-devel python3-pip
+    # requirements
+    - dnf install -y python3-docopt python3-lxml python3-pyyaml python3-pyxattr python3-six python3-future python3-requests
+    # dev-requirements
+    - dnf install -y python3-pytest python3-pytest-xdist python3-pytest-cov python3-mock python3-flake8 python3-sphinx python3-sphinx_rtd_theme.noarch python3-sphinxcontrib-spelling python3-enchant python3-tox make gcc which genisoimage xz
+
+.distro_test: &distro_test
+  script:
+    - python3 setup.py build
+    - export PYTHONPATH=$(pwd)
+    - cd test/unit/ && pytest-3.7 --no-cov-on-fail --cov=kiwi --cov-report=term-missing --cov-fail-under=100 --cov-config .coveragerc -n $(nproc)
+
+tox_37_check:
+  stage: test
+  image: opensuse/tumbleweed
+  before_script:
+    - zypper --non-interactive refresh
+    - zypper --non-interactive install libxml2-devel libxslt-devel gcc python3 python3-devel python3-pip enchant ShellCheck which mkisofs xz gzip
```

Comment:
Fedora ships tox environment for multiple Python versions, so you could have one job template that does it: https://developer.fedoraproject.org/tech/languages/python/multiple-pythons.html

Reply:
I have added tox environments for Python 3.4 and 3.6 using the Fedora containers, but kept the openSUSE containers for Python 2.7 and 3.7.